Sync/BatchingClient.swift

```diff
@@ -0,0 +1,245 @@
+                >>== effect(moveForward) >>> succeed
+        }
+
+        // Send up as a single upload if we're not batching.
+        guard let batch = self.currentBatch else {
+            return self.currentPayload.isEmpty ? succeed() : uploadPayload()
+        }
+
+        // If we have a batch and both it and the payload are empty, don't need to do anything
+        if batch.isEmpty && self.currentPayload.isEmpty {
+            return succeed()
+        }
+
+        if batch.isFull {
+            return commit(batch) >>> uploadPayload
+        } else {
+            return commit(batch) >>> succeed
+        }
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        if records.isEmpty {
+            return succeed()
+        }
+
+        do {
+            var generator = records.generate()
+            while let record = generator.next() {
+                let serializedRecord = try serialize(record)
+                if let uploadOp = accumulateRecord(serializedRecord) {
+                    // Since we couldn't add this record, throw it onto the rest and try adding everything again.
+                    let restOfRecords = generator.map { $0 }
+                    return uploadOp >>> { self.addRecords([record] + restOfRecords) }
+                }
+            }
+        } catch let e {
+            return deferMaybe(e as! MaybeErrorType)
+        }
+        return succeed()
+    }
+
+    private func accumulateRecord(record: UploadRecord) -> DeferredResponse? {
+        // No upload operation happens while queuing records in a payload.
+        if self.currentPayload.add(record) {
+            return nil
+        }
+
+        // Once the payload is full, see if we have a batch we can upload it in. If not, start one.
+        guard var batch = self.currentBatch else {
+            return start()
+        }
+
+        batch.add(self.currentPayload)
+        if batch.isFull {
+            return commit(batch)
+        } else {
+            self.currentBatch = batch
+            return push(batch)
+        }
+    }
+
+    private func serialize(record: Record<T>) throws -> UploadRecord {
+        guard let line = self.serializeRecord(record) else {
+            throw SerializeRecordFailure(record: record)
+        }
+
+        let lineSize = line.utf8.count
+        guard lineSize < Sync15StorageClient.maxRecordSizeBytes else {
+            throw RecordTooLargeError(size: lineSize, guid: record.id)
+        }
+
+        return (record.id, line, lineSize)
+    }
+
+    private func push(batch: Batch) -> DeferredResponse {
+        // Push up the current payload to the server and reset
+        let lines = self.currentPayload.records.map { $0.payload }
+        return uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: batch.batchParams)
+            >>== effect({ _  in self.currentPayload = Payload(config: self.config) })
```

Comment:
Is there a good reason for delaying reassigning `currentPayload` until the upload succeeds? ISTM that you could do this right after computing `lines`, save yourself the memory burden, and avoid the possibility of accidentally adding records to a payload that's already being uploaded.